### ISSUE TITLE
Administrative Locking

### ISSUE THREAD

**Issue by G-Ambatte** [naturalcrit/homebrewery#3326]
### Your idea:

On occasion, the administrators of the Homebrewery website will need to lock a brew down, so that it cannot be accessed by anyone - not even the authors. Specifically, I am thinking of legal action, such as a DMCA takedown request.

My suggestion is that we create the ability to add a new property to the document's stub in the local Homebrewery database, `lock` or something similar.

```
lock: {
  active: true,
  code: 451,
  message: 'This brew has been locked due to DMCA Takedown Request. Please contact the Administrators to resolve this activity.'
}
```

During the `getBrew` process, if the `lock` property exists and `lock.active` is true, an error should be thrown of type `lock.code` with an error message of `lock.message`.

Mock up of final error page:

![image](https://github.com/naturalcrit/homebrewery/assets/19826659/6f2e373f-a8e9-4a36-9713-a62abb5e381a)

I would also expect that a lock page would need to be added to the Admin pages, to allow administrators to apply and remove such locks without needing to access the database directly.

**Comment by ericscheid** [comment]
Would the author still have access to the `/edit/:id` resource?

What if the locked brew is currently being used as a theme source?

**Comment by G-Ambatte** [comment]
As I currently envisage it, the locking would occur during the `getBrew` function, which would prevent access via any method - share, edit, source, download, and presumably also as a theme source.
Theoretically, locking the brew down from any possible method of access should bring the locked status to the authors' attention quickly. In the example of a DMCA takedown request, we would absolutely not want to have the offending material being served to other documents via a shared theme source.

**Comment by 5e-Cleric** [comment]
> Being unable to edit the locked document would make it difficult for the author to fix things.

Hence why there will be a call to message the administrators... although if someone doesn't have discord or reddit this could get weird

Do you think there is worth in creating another gitter room to use as modmail? or simply redirect them to reddit and force them to make an account?

**Comment by ericscheid** [comment]
> > Being unable to edit the locked document would make it difficult for the author to fix things.
> Hence why there will be a call to message the administrators

They will still be unable to make changes, unless the lock is first removed by administrators, necessitating some possibly tricky time coordination (considering timezones etc).

Would perhaps be better then that the editor UI pops an alert, with instructions to the author on how to contact the mods for more information (but still allow editing)

**Comment by calculuschild** [comment]
I think this should block everything except the Edit page, with a popup there as @ericscheid suggests. Don't add any barriers that would make it hard for users to fix their problem.

**Comment by G-Ambatte** [comment]
![image](https://github.com/naturalcrit/homebrewery/assets/19826659/fbdaa7ea-5845-47b9-b6dd-7b26fce91872)
"This is a test message from the database." is the `lock.message` property. All other text in the notification is currently fixed text and will appear in every lock notification.

The "REQUEST LOCK REMOVAL" button is currently attached to a stub function.
![image](https://github.com/naturalcrit/homebrewery/assets/19826659/500d9455-57db-468b-9afb-875e4a1d6f1a)

**Comment by G-Ambatte** [comment]
It occurs to me that we may want two messages - one for the ErrorPage, and one for the LockNotification. The general public needs only to know that the brew is locked, while the author needs to know the specifics of why it is locked.

**Comment by 5e-Cleric** [comment]
> I think this should block everything except the Edit page, with a popup there as @ericscheid suggests. Don't add any barriers that would make it hard for users to fix their problem. Feels way too Kafkaesque.

I agree with this, but, there needs to be some indicator that the brew is locked in edit mode.

**Comment by 5e-Cleric** [comment]
# Mockup for the UI

![image](https://github.com/naturalcrit/homebrewery/assets/91662974/874eb06d-e377-442f-b8a7-d618e7bd8bdc)

**Comment by G-Ambatte** [comment]
As it currently stands, the lock function needs to have options to apply a code, public and private messages.

**Comment by 5e-Cleric** [comment]
As per right now, via #3421 locking works, only with share ID, as intended.

Review is yet to be implemented, and notification in edit page is done in #3382.

I suggest lock state shows up in admin brew lookup function, just to be safe.

**Comment by dbolack-ab** [comment]
@G-Ambatte The last two PRs for this are merged. Is this complete?

**Comment by 5e-Cleric** [comment]
What is left here @G-Ambatte
